### PR TITLE
Use fallbackable path module for node and edge runtime

### PR DESCRIPTION
**Pull request by huozhi**
x-ref: #36190
x-ref: #31506

* Move nodejs ptah module usage to next-server, keep base-server and web-server headless for `'path'`
* Use a native module `path` for nodejs runtime and `path` polyfill for edge runtime

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | huozhi/next.js base-server/remove-path | Change |
| - | - | - | - |
| buildDuration | 19.7s | 19.8s | ⚠️ +109ms |
| buildDurationCached | 7.7s | 7.6s | -99ms |
| nodeModulesSize | 485 MB | 485 MB | ⚠️ +1.11 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | huozhi/next.js base-server/remove-path | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.312 | 4.309 | 0 |
| / avg req/sec | 579.74 | 580.15 | +0.41 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 2.07 | 2.139 | ⚠️ +0.07 |
| /error-in-render avg req/sec | 1207.46 | 1168.78 | ⚠️ -38.68 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js base-server/remove-path | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28.5 kB | 28.5 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 72.1 kB | 72.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js base-server/remove-path | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js base-server/remove-path | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 309 B | 309 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 3.04 kB | 3.04 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.74 kB | 5.74 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.36 kB | 2.36 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 392 B | 392 B | ✓ |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16 kB | 16 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js base-server/remove-path | Change |
| - | - | - | - |
| _buildManifest.js gzip | 461 B | 461 B | ✓ |
| Overall change | 461 B | 461 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js base-server/remove-path | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | huozhi/next.js base-server/remove-path | Change |
| - | - | - | - |
| buildDuration | 23.2s | 23s | -188ms |
| buildDurationCached | 7.6s | 7.6s | -64ms |
| nodeModulesSize | 485 MB | 485 MB | ⚠️ +1.11 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | huozhi/next.js base-server/remove-path | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.39 | 4.342 | -0.05 |
| / avg req/sec | 569.43 | 575.74 | +6.31 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 2.176 | 2.12 | -0.06 |
| /error-in-render avg req/sec | 1149.02 | 1179.15 | +30.13 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js base-server/remove-path | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.9 kB | 28.9 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.8 kB | 72.8 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js base-server/remove-path | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js base-server/remove-path | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 313 B | 313 B | ✓ |
| css-HASH.js gzip | 325 B | 325 B | ✓ |
| dynamic-HASH.js gzip | 3.02 kB | 3.02 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.78 kB | 5.78 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.44 kB | 2.44 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 393 B | 393 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.1 kB | 16.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js base-server/remove-path | Change |
| - | - | - | - |
| _buildManifest.js gzip | 457 B | 457 B | ✓ |
| Overall change | 457 B | 457 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js base-server/remove-path | Change |
| - | - | - | - |
| index.html gzip | 530 B | 530 B | ✓ |
| link.html gzip | 544 B | 544 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: 3da144231558fe2d6bd93bb0adbba3a9376ea954

File: packages/next/lib/constants.ts [hunk 1]
```diff
@@ -1,4 +1,7 @@
-import { join } from 'path'
+import path from '../shared/lib/fallback/path'
+
+const { join } = path
```
Comment (shuding) [inline packages/next/lib/constants.ts:3]:
```suggestion
import { join } from '../shared/lib/with-web-polyfill/path'
```

File: packages/next/server/base-server.ts [hunk 2]
```diff
@@ -182,6 +179,9 @@ export default abstract class Server {
   public readonly hostname?: string
   public readonly port?: number
 
+  protected abstract getDir(_dir: string): string
+  protected abstract getDistDir(): string
+  protected abstract getPagesDir(): string
```
Comment (shuding) [inline packages/next/server/base-server.ts:184]: Are these still necessary to touch?

Comment (huozhi) [reply packages/next/server/base-server.ts:184]: Looks safe to revert 👍

**Review by shuding — APPROVED** [review]

**Pull request merged**